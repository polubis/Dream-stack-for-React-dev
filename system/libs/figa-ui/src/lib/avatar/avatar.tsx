import type { AvatarProps, ImageProps } from './defs';

import c from 'classnames';

const Avatar = ({
  className,
  renderImage,
  size = 'medium',
  shape = 'rounded',
  alt,
  src,
  loading,
  ...props
}: AvatarProps) => {
  const imageProps: ImageProps = {
    alt,
    src,
    loading,
    className: 'avatar-image',
  };

  return (
    <figure className={c('avatar', size, shape, className)} {...props}>
      {renderImage ? (
        renderImage(imageProps)
      ) : (
        // eslint-disable-next-line jsx-a11y/alt-text
        <img {...imageProps} />
      )}
    </figure>
  );
};

export { Avatar };
