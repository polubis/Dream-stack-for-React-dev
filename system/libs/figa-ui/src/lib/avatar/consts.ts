const AVATAR_SIZES = ['small', 'medium', 'big'] as const;
const AVATAR_SHAPES = ['rounded', 'rectangle'] as const;

export { AVATAR_SHAPES, AVATAR_SIZES };
