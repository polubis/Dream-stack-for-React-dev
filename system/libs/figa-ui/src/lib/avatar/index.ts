export * from './avatar';
export * from './defs';
export * from './consts';
