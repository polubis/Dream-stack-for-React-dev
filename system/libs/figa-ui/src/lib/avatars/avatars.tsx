import { Children, cloneElement, useMemo } from 'react';
import type {
  AvatarsProps,
  AvatarsSizeMap,
  MappedChildrenElement,
} from './defs';

import c from 'classnames';
import { Font } from '../font';
import { tokens } from '../theme-provider';

const SIZES: AvatarsSizeMap = {
  small: 400,
  medium: 600,
  big: 800,
};

const pxToNumber = (value: string): number => +value.replace('px', '');

const Avatars = ({
  className,
  children,
  to,
  size = 'medium',
  shape = 'rounded',
  ...props
}: AvatarsProps) => {
  const enhancedChildren = useMemo(() => {
    const avatarSize = tokens.spacing[SIZES[size]];

    const mappedChildren = Children.map(
      children,
      (child: MappedChildrenElement, idx) => (
        <div
          style={{
            transform: `translateX(-${idx * (pxToNumber(avatarSize) / 2)}px)`,
          }}
        >
          {cloneElement(child, { ...child.props, size, shape })}
        </div>
      )
    );

    const slicedChildren = mappedChildren.slice(0, to);
    const restCount = mappedChildren.length - slicedChildren.length;

    if (restCount <= 0) {
      return slicedChildren;
    }

    return [
      ...slicedChildren,
      <div
        className="avatars-rest-count"
        key="rest-count"
        style={{
          height: avatarSize,
          width: avatarSize,
          transform: `translateX(-${
            slicedChildren.length * (pxToNumber(avatarSize) / 2)
          }px)`,
        }}
      >
        <Font variant="b2">+{restCount}</Font>
      </div>,
    ];
  }, [children, to, size, shape]);

  return (
    <div className={c('avatars', className, shape)} {...props}>
      {enhancedChildren}
    </div>
  );
};

export { Avatars };
