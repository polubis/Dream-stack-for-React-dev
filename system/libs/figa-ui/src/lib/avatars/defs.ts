import type { ReactElement, DetailedHTMLProps, HTMLAttributes } from 'react';
import type { AvatarShape, AvatarSize } from '../avatar/defs';
import type { SpacingKey } from '../theme-provider';

type AvatarsHTMLElementProps = Omit<
  DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'children'
>;

type AvatarsProps  = AvatarsHTMLElementProps & {
  children: ReactElement<AvatarsProps> | ReactElement<AvatarsProps>[];
  to: number;
  size?: AvatarSize;
  shape?: AvatarShape;
}

type MappedChildrenElement = ReactElement<AvatarsProps>;

type AvatarsSizeMap = Record<AvatarSize, Extract<SpacingKey, 400 | 600 | 800>>;

export type {
  AvatarsProps,
  AvatarsSizeMap,
  AvatarsHTMLElementProps,
  MappedChildrenElement,
};
