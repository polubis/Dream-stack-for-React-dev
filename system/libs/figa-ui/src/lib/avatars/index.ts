export * from './avatars';
export * from './defs';
