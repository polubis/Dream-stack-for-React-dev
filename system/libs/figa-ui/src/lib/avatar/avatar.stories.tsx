import type { Story, Meta } from '@storybook/react';

import { Avatar } from './avatar';
import { Box } from '../box';
import { AVATAR_SHAPES, AVATAR_SIZES } from './consts';
import { Fragment } from 'react';

export default {
  component: Avatar,
  title: 'Avatar',
} as Meta;

const Template: Story = () => {
  return (
    <Box orientation="row" spacing={AVATAR_SIZES.map(() => 150)}>
      {AVATAR_SIZES.map((size) => (
        <Fragment key={size}>
          {AVATAR_SHAPES.map((shape) => (
            <Box key={shape}>
              <Avatar
                size={size}
                shape={shape}
                alt="My alt text"
                src="https://media-cldnry.s-nbcnews.com/image/upload/rockcms/2022-08/220805-domestic-cat-mjf-1540-382ba2.jpg"
              />
              <Avatar
                size={size}
                shape={shape}
                alt="My alt text"
                src="https://media-cldnry.s-nbcnews.com/image/upload/rockcms/2022-08/220805-domestic-cat-mjf-1540-382ba2.jpg"
                // eslint-disable-next-line jsx-a11y/alt-text
                renderImage={(props) => <img {...props} />}
              />
            </Box>
          ))}
        </Fragment>
      ))}
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
