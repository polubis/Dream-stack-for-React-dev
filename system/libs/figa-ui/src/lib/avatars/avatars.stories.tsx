import type { Story, Meta } from '@storybook/react';

import { Avatars } from './avatars';
import { Avatar } from '../avatar/avatar';
import { Box } from '../box';
import { AVATAR_SHAPES, AVATAR_SIZES } from '../avatar/consts';
import { Font } from '../font';

export default {
  component: Avatars,
  title: 'Avatars',
} as Meta;

const AvatarDummy = (
  <Avatar
    alt="My alt text"
    src="https://media-cldnry.s-nbcnews.com/image/upload/rockcms/2022-08/220805-domestic-cat-mjf-1540-382ba2.jpg"
  />
);

const Template: Story = () => {
  return (
    <Box spacing={[300, 300, 300]} padding={[300, 300, 300, 300]}>
      <Box spacing={[300, 150, 150, 150]}>
        <Font variant="h5">Shapes to use</Font>
        {AVATAR_SHAPES.map((shape) => (
          <Avatars key={shape} shape={shape} to={3}>
            {AvatarDummy}
            {AvatarDummy}
            {AvatarDummy}
            {AvatarDummy}
          </Avatars>
        ))}
      </Box>
      <Box spacing={[300, 150, 150, 150]}>
        <Font variant="h5">Sizes to use</Font>
        {AVATAR_SIZES.map((size) => (
          <Avatars key={size} size={size} to={3}>
            {AvatarDummy}
            {AvatarDummy}
            {AvatarDummy}
            {AvatarDummy}
          </Avatars>
        ))}
      </Box>
      <Box spacing={[300, 150, 150, 150]}>
        <Font variant="h5">Edge cases</Font>
        <Avatars to={3}>
          {AvatarDummy}
          {AvatarDummy}
        </Avatars>
        <Avatars to={1}>
          {AvatarDummy}
          {AvatarDummy}
        </Avatars>
        <Avatars to={1}>{AvatarDummy}</Avatars>
      </Box>
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
