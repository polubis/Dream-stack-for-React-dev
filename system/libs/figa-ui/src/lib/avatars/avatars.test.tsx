import { render, screen } from '@testing-library/react';
import { Avatars } from './avatars';
import { Avatar } from '../avatar/avatar';

describe('Avatars can be used when: ', () => {
  it('[FRAGILE] assigns classes', () => {
    const { container, asFragment } = render(
      <Avatars className="my-class" to={1}>
        <Avatar src="https://" alt="image" />
        <Avatar src="https://" alt="image" />
      </Avatars>
    );

    const component = container.querySelector('.avatars');

    expect(component?.className).toContain('avatars my-class');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders counter badge', () => {
    render(
      <Avatars to={1}>
        <Avatar src="https://" alt="image" />
        <Avatar src="https://" alt="image" />
      </Avatars>
    );

    screen.getByText('+1');
  });

  it('[FRAGILE] calculates offset for children elements', () => {
    const { asFragment } = render(
      <Avatars size="big" to={3}>
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
      </Avatars>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('skips counter badge render if the children count is equal to given limit', () => {
    render(
      <Avatars size="big" to={2}>
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
      </Avatars>
    );

    expect(screen.queryByText('+')).toBeFalsy();
  });

  it('skips counter badge render if the children count is less than given limit', () => {
    render(
      <Avatars size="big" to={1}>
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
      </Avatars>
    );

    expect(screen.queryByText('+')).toBeFalsy();
  });

  it('allows to pass other native HTML element props to container', () => {
    render(
      <Avatars size="big" to={1} title="My title">
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
      </Avatars>
    );

    screen.getByTitle(/My title/);
  });

  it('[FRAGILE] sets default size', () => {
    const { container } = render(
      <Avatars to={1}>
        <Avatar src="https://" alt="image" size="big" />
        <Avatar src="https://" alt="image" size="big" />
      </Avatars>
    );

    const component = container.querySelector('.avatars');

    expect(component?.className).toBe('avatars rounded');
  });

  it('[FRAGILE] injects properties to avatar component', () => {
    const { asFragment } = render(
      <Avatars to={1} size="medium" shape="rectangle">
        <Avatar src="https://" alt="image" size="big" shape="rounded" />
        <Avatar src="https://" alt="image" size="big" shape="rounded" />
      </Avatars>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
