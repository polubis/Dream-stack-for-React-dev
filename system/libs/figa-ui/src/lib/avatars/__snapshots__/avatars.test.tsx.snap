// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatars can be used when:  [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <div
    class="avatars my-class rounded"
  >
    <div
      style="transform: translateX(-0px);"
    >
      <figure
        class="avatar medium rounded"
      >
        <img
          alt="image"
          class="avatar-image"
          src="https://"
        />
      </figure>
    </div>
    <div
      class="avatars-rest-count"
      style="height: 48px; width: 48px; transform: translateX(-24px);"
    >
      <p
        class="font b2 default"
      >
        +1
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatars can be used when:  [FRAGILE] calculates offset for children elements 1`] = `
<DocumentFragment>
  <div
    class="avatars rounded"
  >
    <div
      style="transform: translateX(-0px);"
    >
      <figure
        class="avatar big rounded"
      >
        <img
          alt="image"
          class="avatar-image"
          src="https://"
        />
      </figure>
    </div>
    <div
      style="transform: translateX(-32px);"
    >
      <figure
        class="avatar big rounded"
      >
        <img
          alt="image"
          class="avatar-image"
          src="https://"
        />
      </figure>
    </div>
    <div
      style="transform: translateX(-64px);"
    >
      <figure
        class="avatar big rounded"
      >
        <img
          alt="image"
          class="avatar-image"
          src="https://"
        />
      </figure>
    </div>
    <div
      class="avatars-rest-count"
      style="height: 64px; width: 64px; transform: translateX(-96px);"
    >
      <p
        class="font b2 default"
      >
        +1
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatars can be used when:  [FRAGILE] injects properties to avatar component 1`] = `
<DocumentFragment>
  <div
    class="avatars rectangle"
  >
    <div
      style="transform: translateX(-0px);"
    >
      <figure
        class="avatar medium rectangle"
      >
        <img
          alt="image"
          class="avatar-image"
          src="https://"
        />
      </figure>
    </div>
    <div
      class="avatars-rest-count"
      style="height: 48px; width: 48px; transform: translateX(-24px);"
    >
      <p
        class="font b2 default"
      >
        +1
      </p>
    </div>
  </div>
</DocumentFragment>
`;
